tpp-app/src/services/CalculationService.js

```diff
@@ -0,0 +1,162 @@
+    if (isPeriodOver()) {
+        // Record all period intervals
+        let completeHistory = [];
+
+        // Get all Years
+        GETStoredYears()
+            .then((years) => {
+                years.forEach(async (year) => {
+                    completeHistory.push.apply(completeHistory, await CycleService.GETCycleHistoryByYear(year));
+                });
+            })
+            .catch((e) => {
+                console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+                reject();
+            });
+
+        // Use reduce to find average
+        const averagePeriodLength = completeHistory.reduce(function (sum, interval) {
+            return sum + interval.periodDays;
+        }, 0) / completeHistory.length;
+
+        await AsyncStorage.Set(
+            KEYS.AVERAGE_PERIOD_LENGTH, JSON.stringify(averagePeriodLength)
+        ).then(() => {
+            console.log(`Recalculated ${KEYS.AVERAGE_PERIOD_LENGTH} as ${averagePeriodLength}`)
+            resolve();
+        }).catch((error) => {
+            console.log(`calculateAveragePeriodLength error: ${JSON.stringify(error)}`);
+            reject();
+        });
+    }
+
+    // Else don't calculate, do nothing
+});
+
+/**
+ * Calculates the average cycle length of the user.
+ */
+export const calculateAverageCycleLength = async () => new Promise( async (resolve, reject) => {
+    // Only calculate Average Cycle Length once period is starting and cycle is over
+    if (isPeriodStarting()) {
+        let completeHistory = [];
+
+        // Get all Years
+        GETStoredYears()
+            .then((years) => {
+                years.forEach(async (year) => {
+                    completeHistory.push.apply(completeHistory, await CycleService.GETCycleHistoryByYear(year));
+                });
+            })
+            .catch((e) => {
+                console.log(`calculateAverageCycleLength error: ${JSON.stringify(error)}`);
+                reject();
+            });
+
+        // Use map and reduce to find average
+        const averageCycleLength = completeHistory
+            // Map to difference of days
+            .map((interval, index) => {
+                index === 0 ? 0 : getDaysDiffInclusive(completeHistory[index-1].start, interval.start)
+             })
+            // Remove 0th index
+            .slice(1)
+            // Find average
+            .reduce((sum, cycleDays) => {
+                return sum + cycleDays;
+        }, 0) / completeHistory.length;
+
+        await AsyncStorage.Set(
+            KEYS.AVERAGE_CYCLE_LENGTH, JSON.stringify(averageCycleLength)
+        ).then(() => {
+            console.log(`Recalculated ${KEYS.AVERAGE_CYCLE_LENGTH} as ${averageCycleLength}`)
+            resolve();
+        }).catch((error) => {
+            console.log(`calculateAverageCycleLength error: ${JSON.stringify(error)}`);
+            reject();
+        });
+    }
+
+    // Else don't calculate, do nothing
```

Comment:
Same comment as above. Should `resolve` or `reject` even if it's a "do nothing" case